### PR TITLE
#187300248 admin can assign role to user

### PR DESCRIPTION
**Pull request by PrinceRWIGIMBA**
### What does this PR do?
This PR implements the task related to implementing the Admin Should be able to set roles, It addresses the requirements outlined in Pivotal ID#187300248 .

### Description of Task to be completed?
Admin should be able to assign various users to various roles. There should also be a role for the admin account itself. This will allow only accounts with particular roles permission to perform particular actions.

#### Have the following working

- Admin should be able to create the role
- Admin should be able to assign role to user  
### How should this be manually tested?
1.start your server and add /api/docs to your running port locally
2. test  all role related endpoint  using swagger in browser
### What are the relevant pivotal tracker/Trello stories?
Pivotal ID #187300248.

### CODE REVIEW

**Review by emmanueltct — APPROVED** [review]

**Review by k3lly003 — APPROVED** [review]

File: src/controllers/user.controllers.ts [hunk 1]
```diff
@@ -172,3 +183,63 @@ export const editUser = async (req: any, res: Response) => {
...
   }
 };
+export const assignRoleToUser = async (req: Request, res: Response) => {
+  try {
+    const { roleId } = req.body;
+    const { userId } = req.params;
+
+    const user = await User.findByPk(userId);
+
+    if (!user) {
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: These checks should be done via a middleware.

File: src/controllers/user.controllers.ts [hunk 2]
```diff
@@ -172,3 +183,63 @@ export const editUser = async (req: any, res: Response) => {
...
+      return res.status(404).json({ error: "User not found" });
+    }
+
+    const role = await Role.findByPk(roleId);
+
+    if (!role) {
+      return res.status(404).json({ error: "Role not found" });
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: This should be checked with a middleware.

File: src/controllers/user.controllers.ts [hunk 3]
```diff
@@ -172,3 +183,63 @@ export const editUser = async (req: any, res: Response) => {
...
+  try {
+    const { roleId } = req.params;
+
+    // Find the role by name
+    const role = await Role.findOne({
+      where: { id: roleId }
+    });
+
+    if (!role) {
+      return res.status(404).json({ error: "Role not found" });
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: Use a middleware to check this.

File: src/controllers/user.controllers.ts [hunk 4]
```diff
@@ -15,11 +16,21 @@ export const userSignup = async (req: Request, res: Response) => {
     const { firstName, lastName, email, password } = req.body;
     const userPassword = await passwordEncrypt(password);
 
+    const userRole = await Role.findOne({
+      where: { name: "user" }
+    });
+
+    if (!userRole) {
+      return res.status(500).json({ error: "Default role 'user' not found" });
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: This implementation should be done via middleware.

File: src/controllers/role.controllers.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+export const updateRole = async (req: Request, res: Response) => {
+  try {
+    const { id } = req.params;
+    const { name } = req.body;
+
+    const role = await Role.findByPk(id);
+
+    if (!role) {
+      return res.status(404).json({ error: "Role not found" });
+    }
```
Comment (Eli250) [inline src/controllers/role.controllers.ts:62]: Use a middleware to check this.

File: src/controllers/role.controllers.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+// Delete a role
+export const deleteRole = async (req: Request, res: Response) => {
+  try {
+    const { id } = req.params;
+
+    const role = await Role.findByPk(id);
+
+    if (!role) {
+      return res.status(404).json({ error: "Role not found" });
```
Comment (Eli250) [inline src/controllers/role.controllers.ts:82]: You should use a middleware.

File: src/middlewares/user.middleware.ts [hunk 7]
```diff
@@ -52,4 +54,175 @@ const isValidUserUpdate = (req: Request, res: Response, next: NextFunction) => {
   }
 };
 
-export { isUserExist, isValidUser, isValidUserLogin, isValidUserUpdate };
+const isAdmin = async (req: Request, res: Response, next: NextFunction) => {
+  const authHeader = req.headers.authorization;
+  if (!authHeader) {
+    return res.status(401).json({ error: "Unauthorized, missing token" });
```
Comment (Eli250) [inline src/middlewares/user.middleware.ts]: The authentication middleware should handle this check.

File: src/middlewares/user.middleware.ts [hunk 8]
```diff
@@ -52,4 +54,175 @@ const isValidUserUpdate = (req: Request, res: Response, next: NextFunction) => {
...
+  try {
+    const decodedToken: any = Jwt.verify(
+      token,
+      process.env.JWT_SECRET as string
+    );
+    if (!decodedToken || !decodedToken.email) {
+      return res.status(401).json({ error: "Unauthorized, invalid token" });
+    }
+    // aha ndyi decoder  nshaka gukuramo email
+    const userEmail = decodedToken.email;
```
Comment (Eli250) [inline src/middlewares/user.middleware.ts]: Don't you have authentication middleware?

File: src/middlewares/user.middleware.ts [hunk 9]
```diff
@@ -52,4 +54,175 @@ const isValidUserUpdate = (req: Request, res: Response, next: NextFunction) => {
...
+//     }
+//   } catch (error) {
+//     console.error(error);
+//     return res.status(500).json({ error: "Internal server error" });
+//   }
+// };
+
+// const isBuyer = async (req: Request, res: Response, next: NextFunction) => {
+//   const authHeader = req.headers.authorization;
+//   if (!authHeader) {
```
Comment (Eli250) [inline src/middlewares/user.middleware.ts]: Why all these codes are commented?

File: src/routes/role.routes.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  createRole,
+  getAllRoles,
+  getRoleById,
+  updateRole,
+  deleteRole
+} from "../controllers/role.controllers";
+
+const roleRoutes = express.Router();
+
+roleRoutes.post("/", createRole);
```
Comment (Eli250) [inline src/routes/role.routes.ts]: You are not using any middleware???

File: src/routes/role.routes.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+roleRoutes.get("/", getAllRoles);
```
Comment (Eli250) [inline src/routes/role.routes.ts]: Use middlewares?

**Review by Eli250 — CHANGES_REQUESTED** [review]
You are not using middleware at all. Address all feedback.

**Review by Eli250 — APPROVED** [review]

**Pull request merged**